source/robot_model/src/Model.cpp

```diff
@@ -192,65 +192,78 @@ StateRepresentation::CartesianPose Model::forward_geometry(const StateRepresenta
-    double error;
-
-    bool success = false;
-    Eigen::VectorXd v(this->robot_model_.nv);
-    for (int i = 0;; i++)
-    {
-
-      actual_cartesian_state = this->forward_geometry(actual_joint_state, frame_ids).front();
-      error = desired_cartesian_state.dist(actual_cartesian_state, StateRepresentation::CartesianStateVariable::POSE);
-
-      if (error < eps)
-      {
-        success = true;
-        break;
-      }
-      if (i >= IT_MAX)
-      {
-        success = false;
-        break;
-      }
-      v = this->inverse_kinematic(actual_joint_state, desired_cartesian_state).get_velocities();
-      q = actual_joint_state.get_positions();
-      actual_joint_state.set_velocities(v);
-      StateRepresentation::CartesianTwist x_dot = this->forward_kinematic(actual_joint_state);
-
-      StateRepresentation::CartesianTwist zerovec;
-      zerovec.Zero("zero");
-      double x_dot_norm = x_dot.dist(zerovec, StateRepresentation::CartesianStateVariable::POSE);
-
-      // dt=||x_d - x||/||x'||
-      double dt = error / x_dot_norm;
-
-      q = pinocchio::integrate(this->robot_model_, q, v * dt);
-      actual_joint_state.set_positions(q);
-    }
-    return StateRepresentation::JointPositions(actual_joint_state);
+StateRepresentation::JointPositions Model::inverse_geometry(const StateRepresentation::CartesianState& desired_cartesian_state,
+                                                            const StateRepresentation::JointState& current_joint_state,
+                                                            std::string frame_name, const double& tolerance, const int& max_number_of_iteration) {
+
+  if (frame_name.empty()) {
+    // get last frame if none specified
+    frame_name = this->robot_model_.frames.back().name;
+  }else {
+    // throw error if specified frame does not exist
+    if (!this->robot_model_.existFrame(frame_name)) { throw (Exceptions::FrameNotFoundException(frame_name)); }
   }
 
+  Eigen::VectorXd q = current_joint_state.get_positions();
+
+  const int frame_id = this->robot_model_.getFrameId(frame_name);
+  
+  const pinocchio::SE3 oMdes(desired_cartesian_state.get_orientation().toRotationMatrix(), desired_cartesian_state.get_position());
+
+  double dt;
+  // damping added to the diagonal of J*Jt in order to avoid the singularity
+  const double damp = 1e-6;
+  // if alpha = 1, the Netwton-Raphson method is applied
+  // if 0<alpha <1, the generalized state is iteratively incremented with a persentage of the incrementation provided by Netwton-Raphson 
+  const double alpha = 0.6;
+
+  pinocchio::Data::Matrix6x J(6,this->robot_model_.nq);
+  J.setZero();
+
+  typedef Eigen::Matrix<double, 6, 1> Vector6d;
+  Vector6d err = Vector6d::Ones(6)*100;
+  Eigen::VectorXd v(this->robot_model_.nq);
+
+  int i=0;
+  while((err.norm() > tolerance) && (i < max_number_of_iteration))
+  {
+    pinocchio::framesForwardKinematics(this->robot_model_,this->robot_data_,q);
```

Comment:
I would suggest you try to reuse the functions already implemented in the rest of the class and library in general. For example, you could use the `compute_jacobian` to get the `Jacobian` function and the `forward_geometry` to get the frame pose you are looking for then use the difference between the two `CartesianPose` to compute the error.

Reply:
We don't risk to slow down the algorithm if we add some "structural" complexity?